### PR TITLE
Alpaka implementation allowing device choice at runtime

### PR DESCRIPTION
**Pull request by fwyzard**
Cleaned up version of #13.

### CODE REVIEW

File: main_alpaka.cc [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    for(int i=0; i < NLOOPS; i++){
+    
+      output = std::make_unique<Output>();
+
+      using ViewInput = alpaka::mem::view::ViewPlainPtr<DevHost, Input, Dim, Idx>;
+      ViewInput input_hBuf(&input, devHost, extent);
+
+
+      using BufDevInput = alpaka::mem::buf::Buf<DevAcc, Input, Dim, Idx>;
+      BufDevInput input_dBuf(alpaka::mem::buf::alloc<Input, Idx>(devAcc, extent));
```
Comment (makortel) [inline main_alpaka.cc]: Is the `Buf` essentially a smart-pointer for the device memory? (similar to `unique_ptr` with `cudaMalloc()`+`cudaFree()`)

File: main_alpaka.cc [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+      alpaka::mem::view::copy(queue, input_dBuf, input_hBuf, extent);
```
Comment (makortel) [inline main_alpaka.cc]: If device is CPU, does this function actually copy the data?

File: rawtodigi_alpaka.cc [hunk 3]
```diff
@@ -0,0 +1,527 @@
...
+                                    includeErrors,
+                                    debug
+                                    ));
+
+      alpaka::queue::enqueue(queue, taskRawToDigiKernel);
+
+    }     
+    
+  }; // end namespace alpaka
+}; // end namespace ALPAKA_ARCHITECTURE
```
Comment (makortel) [inline rawtodigi_alpaka.cc]: The inner namespace is `ALPAKA_ARCHITECTURE`, and outer `Alpaka`, no?

(also, please leave remove the redundant semicolon)

**Review by makortel — CHANGES_REQUESTED** [review]

**Comment by fwyzard** [comment]
I think there is some cleanup already in #18 ... I'll move the Alpaka-specific parts here.

File: main_alpaka.cc (hunk 1 above)
Comment (waredjeb) [reply main_alpaka.cc]: Yes, ```Buf``` is a shared pointer for the device memory

File: main_alpaka.cc (hunk 2 above)
Comment (waredjeb) [reply main_alpaka.cc]: As far as I understood, yes it does the copy. But I will check to confirm that

**Comment by fwyzard** [comment]
I've rewritten the commits to avoid touching [rawtodigi_cupla.cc](https://github.com/makortel/pixel-standalone/blob/master/rawtodigi_cupla.cc), and applied some code formatting based on the CMS the clang-format rules.

**Comment by fwyzard** [comment]
Alpaka require separate compilation for different targets, e.g. one for CUDA, one for TBB, one for OpenMP, etc.

As we can see comparing `rawtodigi_alpaka.cc` vs `rawtodigi_alpakaAll.cc`, having a single "main" executable that can choose different backends at runtime require having not only the kernel code, but also the library calls (memory allocation and copies, kernel launch, etc.) to a per-backend file.

@makortel , would you agree to make a similar change for all implementations ?

For example, we could have
  - one "rawtodigi_xxx.cc" file with the device code and kernel
  - one (for lack of a better name) "analyzer_xxx.cc" file that handles the memory operations and kernel launches 
  - one "main_xxx.cc" file with the functionality to read the input, run the loop and print the statistics

**Review by makortel — APPROVED** [review]

**Pull request merged**